File: csharp/EcoStatCollectionBuilder.cs
using System.Collections.Generic;

public class EcoStatCollectionBuilder
{
    private List<EcoStat> defaultEcoStatCollection =
        new List<EcoStat>(){
            new EcoStat("David Allen", 1.234f),
            new EcoStat("Lisa Crispin", 0.025f),
            new EcoStat("Ian Dees", 98.76f)
        };

    private Dictionary<string, float> ecoStatData = null;

    public List<EcoStat> Build()
    {
        if (ecoStatData == null)
        {
            return defaultEcoStatCollection;
        }
        else
        {
            List<EcoStat> ecoStatCollection = new List<EcoStat>();

            foreach(KeyValuePair<string, float> entry in ecoStatData)
            {
                ecoStatCollection.Add(new EcoStat(entry.Key, entry.Value));
            }

            return ecoStatCollection;
        }
    }

    public EcoStatCollectionBuilder WithSalesPersonRpm(string name, float rpm)
    {
        if (ecoStatData == null)
        {
            ecoStatData = new Dictionary<string, float>();
        }

        ecoStatData[name] = rpm;
        return this;
    }
}

File: cpp/eco_stat_collection_builder.cpp
#include "eco_stat_collection_builder.hpp"

namespace shouty_report
{
  const std::vector<eco_stat> eco_stat_collection_builder::default_eco_stat_collection= {
     eco_stat("David Allen", 1.234f),
     eco_stat("Lisa Crispin", 0.025f),
     eco_stat("Ian Dees", 98.76f)
 };

  std::vector<eco_stat> eco_stat_collection_builder::build() const
  {
      if (eco_stat_data.size() == 0)
      {
          return default_eco_stat_collection;
      }
      else
      {
          std::vector<eco_stat> eco_stat_collection;

          for (auto & entry : eco_stat_data)
          {
              eco_stat_collection.push_back(
                  eco_stat(entry.first, entry.second)
              );
          }

          return eco_stat_collection;
      }
  }

  eco_stat_collection_builder &
  eco_stat_collection_builder::with_sales_person_rpm(
      const std::string & name,
      double rpm
  )
  {
      eco_stat_data[name] = rpm;

      return *this;
  }

}
